### ISSUE TITLE
Inconsistent handling of directives in buildASTSchema and buildClientSchema

### ISSUE THREAD

**Issue by Cito** [graphql/graphql-js#3419]
I noticed a difference in how specified (standard) directives are added by `buildASTSchema` vs `buildClientSchema`.

In the former function, the specified directives [are added](https://github.com/graphql/graphql-js/blob/90bd6ff72625173dd39a1f82cfad9336cfad8f65/src/utilities/buildASTSchema.ts#L83) automatically if they are missing.

In the latter function, the specified directives [are not added](https://github.com/graphql/graphql-js/blob/90bd6ff72625173dd39a1f82cfad9336cfad8f65/src/utilities/buildClientSchema.ts#L105) automatically if they are missing.

I think this should be handled in the same way by both method - this different behavior caused some head-scratching [here](https://github.com/graphql-python/gql/issues/278).

**Comment by IvanGoncharov** [comment]
Also, I expected `GraphQLSchema => GraphQLExecutableSchema` conversion to be implemented in GraphQL servers so user code and an absolute majority of schema generating/transforming libraries to be unaffected.
Another benefit for shifting the injection of directives closer to `execute` (basically inside GraphQL servers) is that it allows servers to make a decision on supporting or not supporting directive-based features (e.g. `@stream/@defer`).

**Comment by IvanGoncharov** [comment]
BTW. We also have inconsistency with assigning root type. For example:
```
const schema = buildSchema(`
  type Query {
    foo: String
  }
`);

const fullSchema = extendSchema(schema, parse(`
  type QueryRoot {
    query: Query
  }
  
  schema {
    query: QueryRoot
    # ^ error here because we already have `Query` as query root type
  }
`);
```
But if you pass it as single SDL everything works fine:
```
const schema = buildSchema(`
  type Query {
    foo: String
  }

  type QueryRoot {
    query: Query
  }
  
  schema {
    query: QueryRoot
    # ^ error here because we already have `Query` as query root type
  }
`);
```
My proposal is to have `GraphQLExecutableSchema` as a single point to do all non-trivial logic/defaults (including making `Query` a root type).

**Comment by yaacovCR** [comment]
> Also, I expected `GraphQLSchema => GraphQLExecutableSchema` conversion to be implemented in GraphQL servers so user code and an absolute majority of schema generating/transforming libraries to be unaffected.
> Another benefit for shifting the injection of directives closer to `execute` (basically inside GraphQL servers) is that it allows servers to make a decision on supporting or not supporting directive-based features (e.g. `@stream/@defer`).

I agree, this should be in user code, and it could be class based, or a simple function. A new type is not strictly necessary, but may be beneficial. This is basically similar to the path taken by graphql executor where we assume users will call validateSchema as a separate step when necessary.

**Comment by yaacovCR** [comment]
> BTW. We also have inconsistency with assigning root type. For example:
> 
> ```
> const schema = buildSchema(`
>   type Query {
>     foo: String
>   }
> `);
> 
> const fullSchema = extendSchema(schema, parse(`
>   type QueryRoot {
>     query: Query
>   }
>   
>   schema {
>     query: QueryRoot
>     # ^ error here because we already have `Query` as query root type
>   }
> `);
> ```
> 
> But if you pass it as single SDL everything works fine:
> 
> ```
> const schema = buildSchema(`
>   type Query {
>     foo: String
>   }
> 
>   type QueryRoot {
>     query: Query
>   }
>   
>   schema {
>     query: QueryRoot
>     # ^ error here because we already have `Query` as query root type
>   }
> `);
> ```

The original question was about inconsistencies between `buildASTSchema` and `buildClientSchema`, but this supposes an inconsistency between passing a single SDL string to `buildSchema` (really eventually `builtASTSchema`) versus two strings, with the two strings, of course, using `extendSchema`. This "inconsistency" is really because the second string is not an "extension," it's a transformation, switching the query root type from 'Query' to 'RootQuery'. What one would near here is a transformSchema (like graphql-tools mapSchema) and presumably a `transform` keyword instead of the `extend` schema. This is a really interesting idea, but I don't see how it's directly related.

> My proposal is to have `GraphQLExecutableSchema` as a single point to do all non-trivial logic/defaults (including making `Query` a root type).

I read this as you possibly suggesting that within the core GraphQLSchema, the Query root type would not be defined at all (possibly never defined, possibly just not set up as "Query" by default, and that the userland GraphQLExecutableSchema class would be responsible for this? Or is that a typo, and you meant to say the following:

> My proposal is to have `GraphQLExecutableSchema` as a single point to do all non-trivial logic/defaults (including making `QueryRoot` a root type).

This would make sense to me, that any non trivial transformation logic should be in userland, whether it's in a new class or a new function. Are you able to clarify? Thanks!
